### ISSUE TITLE
[Inductor] Fusion of Tiled Point-Wise and Reduction Operators

### ISSUE THREAD

**Issue by y-sq** [pytorch/pytorch#128063]
### 🚀 The feature, motivation and pitch

Repro of the issue:
```
    def test(x):
        y = torch.sum(x)
        z = x / 10.0
        z_t = z.t().contiguous().t()
        return y, z, z_t
```

We expect the first-level reduction of `sum` and `(x + x.t())` to be fused into one operator. 

However, these two ops currently can't fuse due to `invalid tiling`. Assuming the shape of x is 4096x4096. In inductor the first reduction will be split to 2 layers to be more efficient. First layer reduce [4096 x 4096] to [X] and the second layer reduce [X] to [1]. When inductor decide X, it does not realize that picking 4096 would make it possible to fuse the first layer reduction with the following pointwise op. In this case, inductor picks X other than 4096 (512 here). 

cc @ezyang @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @yanbing-j @vkuzo @albanD @msaroufim @bdhirsh

**Comment by peterbell10** [comment]
I tried manually splitting the reduction in python:
```python
def test(x):
    y = torch.sum(x, dim=1)
    z = (x + x.t()) / 10.0 # x need to be a squared tensor here
    return y.sum(), z
```
And indeed it does allow the fusion to happen and the fused kernels are faster.

**Comment by vkuzo** [comment]
I still repro this test case with the same fusion log output.  Furthermore, on the latest PyTorch nightly + torchao, I see the same result (reduction not fused with pointwise), but the fusion log output does not show "invalid tiling":

```
> TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TORCH_LOGS="fusion" TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 python benchmarks/float8/profile_linear_float8.py ~/local/tmp/profile_20241030 --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed --add_inductor_metadata_to_trace False
...
V1030 14:03:01.056000 2462143 site-packages/torch/_inductor/scheduler.py:2305] [1/0] [__fusion] ===== attempting fusion (1/10): 19 nodes =====                                                                                            
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2594] [1/0] [__fusion] fuse_nodes_once, candidates:                                                                                                              
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op0'), Reduction(['[32768]', 'max', 'origins=OrderedSet([max_1, abs_1])'])                                          
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op1'), Reduction(['[512]', 'max', 'origins=OrderedSet([max_1, abs_1])'])                                            
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op2'), Pointwise(['[]', 'origins=OrderedSet([copy_1, copy__default])'])                                             
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op3'), Reduction(['[16]', 'max', 'origins=OrderedSet([max_2])'])                                                    
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op4'), Reduction(['[32768]', 'max', 'origins=OrderedSet([max_4, abs_3])'])                                          
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op5'), Reduction(['[512]', 'max', 'origins=OrderedSet([max_4, abs_3])'])                                            
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op6'), Pointwise(['[]', 'origins=OrderedSet([copy_5, copy__default_1])'])                                           
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op7'), Reduction(['[16]', 'max', 'origins=OrderedSet([max_5])'])                                                    
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op8'), Pointwise(['[1]', 'origins=OrderedSet([convert_element_type_1, clamp_min, reciprocal_2, reciprocal, copy_2, mul, convert_element_type])'])                                                                                                                                                                                                             
V1030 14:03:01.057000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op9'), Pointwise(['[1]', 'origins=OrderedSet([clamp_min_2, reciprocal_1, convert_element_type_4, copy_6, mul_2, convert_element_type_5, reciprocal_3])'])                                                                                                                                                                                                     
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op10'), Pointwise(['[4096, 4096]', 'origins=OrderedSet([clamp_min, view, convert_element_type_2, reciprocal, mul, mul_1, convert_element_type_1, _scaled_mm, clamp_min_1, clamp_max, copy_2, convert_element_type, convert_element_type_3])'])                                                                                                                
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op11'), Pointwise(['[4096, 4096]', 'origins=OrderedSet([clamp_min, view, convert_element_type_2, reciprocal, mul, mu
l_1, convert_element_type_1, _scaled_mm, clamp_min_1, clamp_max, copy_2, convert_element_type, convert_element_type_3])'])                                                                                                                
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   ExternKernelSchedulerNode(name='op12')                                                                                                  
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op13'), Pointwise(['[4096, 4096]', 'origins=OrderedSet([clone])'])                                                  
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op14'), Pointwise(['[4096, 4096]', 'origins=OrderedSet([clone_2])'])                                                
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op15'), Pointwise(['[1]', 'origins=OrderedSet([copy_3, copy, copy__3])'])                                           
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op16'), Pointwise(['[1]', 'origins=OrderedSet([convert_element_type_1, clamp_min, copy__5, reciprocal, copy_2, mul, 
convert_element_type])'])                                                                                                                                                                                                                 
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op17'), Pointwise(['[1]', 'origins=OrderedSet([copy__6, copy_7, copy_4])'])                                         
V1030 14:03:01.058000 2462143 site-packages/torch/_inductor/scheduler.py:2596] [1/0] [__fusion]   SchedulerNode(name='op18'), Pointwise(['[1]', 'origins=OrderedSet([copy__8, reciprocal_1, convert_element_type_4, copy_6, mul_2, convert
_element_type_5, clamp_min_2])'])                                                                                                                                                                                                         
V1030 14:03:01.060000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op0 with op10: no shared data                                                                                                  
V1030 14:03:01.061000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op0 with op14: no shared data                                                                                                  
V1030 14:03:01.061000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op10 with op14: score_fusion_memory_threshold                                                                                  
V1030 14:03:01.062000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op0 with op1: numel/rnumel mismatch (reduce) (512, 1), (32768, 512)                                                            
V1030 14:03:01.062000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op2 with op15: score_fusion_memory_threshold                                                                                   
V1030 14:03:01.062000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op2 with op3: no shared data                                                                                                   
V1030 14:03:01.062000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op10 with op3: nodes numel incompatibility
V1030 14:03:01.062000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op14 with op3: nodes numel incompatibility
V1030 14:03:01.062000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op8 with op10: score_fusion_memory_threshold
V1030 14:03:01.063000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op8 with op14: score_fusion_memory_threshold
V1030 14:03:01.063000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op8 with op16: score_fusion_memory_threshold
V1030 14:03:01.063000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op10 with op16: score_fusion_memory_threshold
V1030 14:03:01.063000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op14 with op16: score_fusion_memory_threshold
V1030 14:03:01.063000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op4 with op11: no shared data
V1030 14:03:01.064000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op4 with op13: no shared data
V1030 14:03:01.064000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op11 with op13: score_fusion_memory_threshold
V1030 14:03:01.064000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op4 with op5: numel/rnumel mismatch (reduce) (512, 1), (32768, 512)
V1030 14:03:01.064000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op6 with op17: score_fusion_memory_threshold
V1030 14:03:01.065000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op6 with op7: no shared data
V1030 14:03:01.065000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op11 with op7: nodes numel incompatibility
V1030 14:03:01.065000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op13 with op7: nodes numel incompatibility
V1030 14:03:01.065000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op9 with op11: score_fusion_memory_threshold
V1030 14:03:01.065000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op9 with op13: score_fusion_memory_threshold
V1030 14:03:01.065000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op9 with op18: score_fusion_memory_threshold
V1030 14:03:01.066000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op11 with op18: score_fusion_memory_threshold
V1030 14:03:01.066000 2462143 site-packages/torch/_inductor/scheduler.py:742] [1/0] [__fusion] cannot fuse op13 with op18: score_fusion_memory_threshold
V1030 14:03:01.066000 2462143 site-packages/torch/_inductor/scheduler.py:2722] [1/0] [__fusion] found 8 possible fusions
V1030 14:03:01.066000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op7 with op9
V1030 14:03:01.066000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op3 with op8
V1030 14:03:01.066000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op7_op9 with op18
V1030 14:03:01.067000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op3_op8 with op16
V1030 14:03:01.067000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op1 with op2
V1030 14:03:01.067000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op5 with op6
V1030 14:03:01.067000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op5_op6 with op17
V1030 14:03:01.067000 2462143 site-packages/torch/_inductor/scheduler.py:2605] [1/0] [__fusion] fusing op1_op2 with op15
V1030 14:03:01.067000 2462143 site-packages/torch/_inductor/scheduler.py:2312] [1/0] [__fusion] completed fusion round (1/10): fused 19 nodes into 11 nodes

```

full output: https://gist.github.com/vkuzo/28883d486c3e8147322f742be2492548

**Comment by vkuzo** [comment]
setting a breakpoint in the scheduler, I see the following printout of the two nodes' reads:

```
> /data/users/vasiliy/pytorch/torch/_inductor/scheduler.py(3038)can_fuse()
-> why("no shared data")
(Pdb) node1
SchedulerNode(name='op0')
(Pdb) node2
SchedulerNode(name='op10')
(Pdb) shared_data_score
0
(Pdb) node1.read_writes
ReadWrites(reads=OrderedSet([MemoryDep('primals_1', 32768*d0 + d1, {d0: 512, d1: 32768}, None)]), writes=OrderedSet([MemoryDep('buf0', d0, {d0: 512}, None)]), index_exprs=OrderedSet([]), range_vars=[d0, d1], var_ranges={d0: 512, d1: 32768})
(Pdb) node2.read_writes
ReadWrites(reads=OrderedSet([MemoryDep('primals_1', 4096*d0 + d1, {d0: 4096, d1: 4096}, None), MemoryDep('buf3', 0, {}, None)]), writes=OrderedSet([MemoryDep('buf10', 4096*d0 + d1, {d0: 4096, d1: 4096}, None)]), index_exprs=OrderedSet([]), range_vars=[d0, d1], var_ranges={d0: 4096, d1: 4096})
```

So, node1 and node2 are both reading from primals_1, but with different block sizes.  It seems very similar to the reproduction posted by @y-sq at the top of the issue.

**Comment by y-sq** [comment]
@vkuzo will you see "cannot fuse, invalid tiling" is not enabling "TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1"?

My guess is "TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1" reorders how the input tensor of fused_cast_transposed is accessed (such as, from "first loop each row", to "first loop each column / loop it as a transposed tensor". When Inductor checks whether two ops share the same data, it checks both the tensor itself and how the tensor is accessed. If they don't match, it's "no shared data"/
We might need something loop reordering here for reduction. (Shunting's previous change only enables it to pointwise for now; You may find some additional context in our group chat of "float8 + inductor".)

**Comment by vkuzo** [comment]
if I disable loop ordering after fusion, I then see `cannot fuse op0 with op10: intermediate nodes between node1 & node2`

**Comment by vkuzo** [comment]
AFAIK this has not been fixed and we still need a solution.  @eellison , is cooperative reductions still being planned for this year / still expected to help with this?
